plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Stablecoin.hs

```diff
@@ -350,7 +399,29 @@ contract = do
     let theClient = machineClient (scriptInstance sc) sc
     _ <- mapError StateMachineError $ StateMachine.runInitialise theClient (initialState theClient) mempty
     forever $ do
-        mapError RunStepError (endpoint @"run step") >>= mapError StateMachineError . StateMachine.runStep theClient
+        i <- mapError RunStepError (endpoint @"run step")
+        checkTransition theClient sc i
+        mapError StateMachineError $ StateMachine.runStep theClient i
+
+-- | Apply 'checkValidState' to the states before and after a transition
+--   and log a warning if something isn't right.
+checkTransition :: StateMachineClient BankState Input -> Stablecoin -> Input -> Contract StablecoinSchema StablecoinError ()
+checkTransition theClient sc i@Input{inpConversionRate} = do
+        currentState <- mapError StateMachineError $ StateMachine.getOnChainState theClient
+        case checkHashOffChain inpConversionRate of
+            Right Observation{obsValue} -> do
+                case currentState of
+                    Just ((TypedScriptTxOut{tyTxOutData}, _), _) -> do
+                        case checkValidState sc tyTxOutData obsValue of
+                            Right _ -> logInfo @String "Current state OK"
+                            Left w  -> logWarn $ "Current state is invalid: " <> show w
```

Comment:
Just checking: this can happen when the observed rate has changed such that the reserves are inadequate, right? Is there anything that can be done in that case currently?

Reply:
Yes correct. If that happens then the client prints out a warning, but the transition can still be valid, as long as the state is valid again after the transition. So you just need to make sure the reserve ratio is in the allowed bounds after the transition.

Reply:
Then I guess what I really want is to be clear about what it means for the state to be invalid: in particular it *doesn't* mean your machine can't progress or is stuck! Whereas if the state you transition *to* is invalid then we do block it. It's just a bit confusing.

Reply:
Ok, I'll change the log output to make that clearer.